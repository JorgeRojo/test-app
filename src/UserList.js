import React, { useEffect, useState, useCallback } from 'react';
import get from 'lodash.get';
import Paginator from './Paginator';
import { wrap } from 'module';

const UserList = () => {
  const [isError, setIsError] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const [responseData, setResponseData] = useState(null);
  const [page, setPage] = useState(1);

  useEffect(() => {
    (async () => {
      setIsLoading(true);
      try {
        const response = await fetch(
          `https://reqres.in/api/users?page=${page}`
        );
        if (response.status === 200) {
          setResponseData(await response.json());
          setIsError(false);
        } else {
          setIsError(true);
        }
      } catch (error) {
        setResponseData(null);
        setIsError(true);
      }
      setIsLoading(false);
    })();
  }, [page]);

  const handlePageChange = useCallback(setPage, []);

  if (isLoading) {
    return <div>Loading...</div>;
  }

  if (isError) {
    return <div>Error loading!</div>;
  }

  const data = get(responseData, 'data', []);

  return (
    <>
      <div style={styles.cardList}>
        {data.map(({ email, first_name, last_name, avatar }, index) => (
          <div key={index} style={styles.card}>
            <img width="80" src={avatar} alt={first_name} />
            <h2>
              {first_name} {last_name}
            </h2>
            <a href={`mailto:${email}`}>{email}</a>
          </div>
        ))}
      </div>
      <Paginator {...{ responseData, handlePageChange }} />
    </>
  );
};

const styles = {
  cardList: { display: 'flex', flexWrap: 'wrap' },
  card: { border: 'solid 1px #ccc', padding: '15px', margin: '15px' },
};

export default UserList;
