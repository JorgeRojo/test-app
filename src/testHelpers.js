import React from 'react';

export const mockComponent = (name, { children, ...props }) => (
  <code name={name} props={JSON.stringify(props, null, 2)}>
    {children}
  </code>
);
