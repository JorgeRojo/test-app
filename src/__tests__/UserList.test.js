import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { mockComponent } from '../testHelpers';
import UserList from '../UserList';

const mockResponseData = {
  page: 2,
  per_page: 3,
  total: 12,
  total_pages: 4,
  data: [
    {
      id: 4,
      email: 'eve.holt@reqres.in',
      first_name: 'Eve',
      last_name: 'Holt',
      avatar:
        'https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg',
    },
    {
      id: 5,
      email: 'charles.morris@reqres.in',
      first_name: 'Charles',
      last_name: 'Morris',
      avatar:
        'https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg',
    },
    {
      id: 6,
      email: 'tracey.ramos@reqres.in',
      first_name: 'Tracey',
      last_name: 'Ramos',
      avatar:
        'https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg',
    },
  ],
};

jest.mock('../Paginator', () => props => mockComponent('Paginator', props));

describe('UserList.js', () => {
  let container = null;

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('renders without crashing', () => {
    render(<UserList />, container);
    unmountComponentAtNode(container);
  });

  it('renders loader before data is loaded', async () => {
    act(() => {
      render(<UserList />, container);
    });
    expect(container).toMatchSnapshot();
    expect(container.textContent).toBe('Loading...');
  });

  it('renders error when api call fails', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.reject({
        status: 500,
      })
    );
    await act(async () => {
      render(<UserList />, container);
    });
    expect(container).toMatchSnapshot();
    expect(container.textContent).toBe('Error loading!');
    global.fetch.mockRestore();
  });

  it('renders error when api call is wrong', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        status: 404,
        json: () => Promise.resolve(mockResponseData),
      })
    );
    await act(async () => {
      render(<UserList />, container);
    });
    expect(container).toMatchSnapshot();
    expect(container.textContent).toBe('Error loading!');
    global.fetch.mockRestore();
  });

  it('renders users data with success api call', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve(mockResponseData),
      })
    );
    await act(async () => {
      render(<UserList />, container);
    });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(
      'https://reqres.in/api/users?page=1'
    );
    expect(container).toMatchSnapshot();
    global.fetch.mockRestore();
  });
});
