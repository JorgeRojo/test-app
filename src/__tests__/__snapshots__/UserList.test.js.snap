// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList.js renders error when api call fails 1`] = `
<div>
  <div>
    Error loading!
  </div>
</div>
`;

exports[`UserList.js renders error when api call is wrong 1`] = `
<div>
  <div>
    Error loading!
  </div>
</div>
`;

exports[`UserList.js renders loader before data is loaded 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`UserList.js renders users data with success api call 1`] = `
<div>
  <div
    style="display: flex; flex-wrap: wrap;"
  >
    <div
      style="border: 1px solid #ccc; padding: 15px; margin: 15px;"
    >
      <img
        alt="Eve"
        src="https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        width="80"
      />
      <h2>
        Eve
         
        Holt
      </h2>
      <a
        href="mailto:eve.holt@reqres.in"
      >
        eve.holt@reqres.in
      </a>
    </div>
    <div
      style="border: 1px solid #ccc; padding: 15px; margin: 15px;"
    >
      <img
        alt="Charles"
        src="https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
        width="80"
      />
      <h2>
        Charles
         
        Morris
      </h2>
      <a
        href="mailto:charles.morris@reqres.in"
      >
        charles.morris@reqres.in
      </a>
    </div>
    <div
      style="border: 1px solid #ccc; padding: 15px; margin: 15px;"
    >
      <img
        alt="Tracey"
        src="https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
        width="80"
      />
      <h2>
        Tracey
         
        Ramos
      </h2>
      <a
        href="mailto:tracey.ramos@reqres.in"
      >
        tracey.ramos@reqres.in
      </a>
    </div>
  </div>
  <code
    name="Paginator"
    props="{
  \\"responseData\\": {
    \\"page\\": 2,
    \\"per_page\\": 3,
    \\"total\\": 12,
    \\"total_pages\\": 4,
    \\"data\\": [
      {
        \\"id\\": 4,
        \\"email\\": \\"eve.holt@reqres.in\\",
        \\"first_name\\": \\"Eve\\",
        \\"last_name\\": \\"Holt\\",
        \\"avatar\\": \\"https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg\\"
      },
      {
        \\"id\\": 5,
        \\"email\\": \\"charles.morris@reqres.in\\",
        \\"first_name\\": \\"Charles\\",
        \\"last_name\\": \\"Morris\\",
        \\"avatar\\": \\"https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg\\"
      },
      {
        \\"id\\": 6,
        \\"email\\": \\"tracey.ramos@reqres.in\\",
        \\"first_name\\": \\"Tracey\\",
        \\"last_name\\": \\"Ramos\\",
        \\"avatar\\": \\"https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg\\"
      }
    ]
  }
}"
  />
</div>
`;
